### ISSUE TITLE
feature request: allow user to set individual line types

### ISSUE THREAD

**Issue by joshhjacobson** [BigFatDog/parcoords-es#40]
Hi again,

In the same way that a user can specify `pc.color()` with the argument being either a single color or a function, I think it would be useful if they could also specify line type. Canvas has the `lineWidth` property and `setLineDash()` method. Using these tools in conjunction, the user could specify many different line types. It could look something like this:

```javascript
pc.lineType(width = 2, dash = [5, 15])
```
or
```javascript
pc.lineType(function)
```

For charts with lots of lines this could get messy, but for more simple charts I think this could be very powerful. What are your thoughts? Would this be straightforward to implement?

Best,
Josh

**Comment by BigFatDog** [comment]
Hi Josh,
Interesting. I agree to make `lineWidth` and `highlightedLineWidth` configurable. But I need a second thought on making `lineStyle` configurable. 

# Approach 1: A general method for all lines
If the lineType is per datum (a conventional model for most of parcoords' apis), client's usage would be:
```
// client's usage
pc.lineType( (d, i) => {
    if (d.isMarked === true) {
         return { width: 5, style: 'solid' };
    } else if (d.isHighlighed === true) {
         return {  width: 3, style: 'solid'  };
    } else {
         return {  width: 1, style: 'dashed'  };
    }
})
```
It would be a burden on end users because:
1. this api requires users know too many things
2. We designed api separately, e.g., `color`, `brushedColor` and `markedShadowColor`. This api breaks the convention.

# Approach 2: 6 apis  for regular lines, highlighted ones and marked ones
```
// line width
pc.lineWidth(5);
pc.highlightedLineWidth(3);
pc.markedLineWidth(3); // you have already made this one.

// when parameter is a function:
const lineWidth = (d, i) => d.group === 'sample_group' ? 5 : 3);

// similarly for lineStyle:
pc.lineStyle(); 
pc.highlightedLineStyle();
pc.markedLineStyle();

// when parameter is object
pc.lineStyle = (style = 'solid', segments = [5, 15]) => ...
// segments is taken into account only when style is 'dashed';
// when parameter is a function:
const style = (d, i) => d.group === 'sample_group' ? { style: 'solid'} : { style: 'dashed': segments: [5, 10]}
pc.lineStye(style);
```
Disadvantage of this approach is it might be over engineering... 🤔

**Comment by joshhjacobson** [comment]
Hmmm, you raise some very good points! I would be inclined to go with the following:
```javascript
// general method, unspecified parameters & arguments (i.e. highlighted) set to default
pc.lineType( marked = { width: 5, style: 'dashed', segments: [5, 15] }, 
             general = { width: 3 } ) 
```
But like you mentioned, this breaks convention.

Alternatively, the second approach leaves us with:
```javascript
pc.lineWidth()
pc.highlightedLineWidth()
pc.markedLineWidth()

pc.lineStyle()
pc.highlightedLineStyle()
pc.markedLineStyle()
```
and maybe also
```javascript
pc.markedColor(object or function)
```
This does seem like a lot, so I guess we have to weigh our options between freedom of flexibility and easy of use. For reference, the folks at [ggplot](https://ggplot2.tidyverse.org/reference/) seem to be fine with a verbose (and even repetitive) api in the name of flexibility.

**Comment by BigFatDog** [comment]
Yes, it is over complicated. I tend to not to implement this feature for now. I hope it is OK for you. 

I've have a look at the `ggplot`. I will have a closer look at it, not for this feature, but for my future projects.

Thanks for this feature request.

**Comment by NateSavidge** [comment]
Hi everyone,

I've just migrated my project from the original syntagmatic ParCoords library to this one. Thanks for all of the work you have put into this project!

One of the features that I had added to the original library was the ability to set a custom color for objects that are highlighted. When a line is marked, and then a second line is highlighted from another data viz component (like a data table for example), it's difficult to identify which line is which.

It would be really useful to have API access to the settings you discussed above - something like the "Approach 2". Is there any interest in revisiting this request?

If not, I can just follow the pattern established in the lineWidth and add `highlightedLineColor` to the config settings.
